### ISSUE TITLE
opened module environment is not right-away available

### ISSUE THREAD

**Issue by ulugbekna** [ocaml/merlin#1135]
**Problem:** The environment is not correctly extended with the opened module before some expression is used. 

When `Base` is opened, there must be more completions to `String.to_<...>` as you can see in the second screenshot.

Screenshot 1 
![image](https://user-images.githubusercontent.com/16353531/80310264-0ed9da80-87da-11ea-98a1-62878415b723.png)

Putting some expression, e.g., `let () = ();;`, fixes the environment.

Screenshot 2
![image](https://user-images.githubusercontent.com/16353531/80310277-1a2d0600-87da-11ea-926d-d38dbcedc9aa.png)

Screenshot 3
![image](https://user-images.githubusercontent.com/16353531/80310599-e357ef80-87db-11ea-9b50-5bb7f2836ddf.png)

Since it happens on VS Code with both extensions: OCaml Platform, which uses ocaml-lsp, and Reason and OCaml IDE, which works directly with merlin, I assumed it's merlin problem.

**Merlin version:**
I am using latest merlin version (`ocamlmerlin -version` gives `The Merlin toolkit version 73e9771b, for Ocaml 4.09.1`). 

Sorry if it was reported before.

**Comment by copy** [comment]
Possibly related to https://github.com/ocaml/merlin/issues/774

**Comment by trefis** [comment]
Indeed, this is the same issue as #774 , and it's been a long standing one.

I thought we had fixed it at some point, but apparently not (or it crept back in).

**Comment by ulugbekna** [comment]
Maybe helpful: the problem is only with toplevel opens, i.e., 

```ocaml
let _ = 
  let open Base in
  String.to_<|>
```

works as expected

**Comment by let-def** [comment]
It is not only open, it is all top-level definitions.
The problem has to do with how merlin selects the environment when the definition under the cursor is not parseable. It fall backs to the previous one and therefore does not see an environment where this previous definition has been registered.

**Comment by ulugbekna** [comment]
Ah, then could we do: when the expr under the cursor is not parseable, if the previous expression modifies the environment, the new environment should be used as default for the non-parseable expr under cursor?

Also, how is it that the example with `let open` in my latest comment works fine? Are let defs treated differently than top-level `open`s? :-D

**Comment by copy** [comment]
This seems to be fixed in most cases.

```ocaml
open Base
let _ = String.<complete> (* works *)
```

```ocaml
module Foo = struct (* ... *) end;;
Foo.<complete> (* works *)
```

```ocaml
open Base;;
String.<complete> (* still doesn't work *)
```

**Comment by voodoos** [comment]
> This seems to be fixed in most cases.
> 
> ```ocaml
> open Base
> let _ = String.<complete> (* works *)
> ```
> 
> ```ocaml
> module Foo = struct (* ... *) end;;
> Foo.<complete> (* works *)
> ```
> 
> ```ocaml
> open Base;;
> String.<complete> (* still doesn't work *)
> ```

if I'm not mistaken the last problematic case is exactly the one described in the initial issue.
